ontoconv/pipelines.py

```diff
@@ -183,6 +196,155 @@ def generate_pipeline(
+
+    Arguments:
+        ts: Tripper triplestore documenting data sources and sinks.
+        resources: A dict referring to OTEAPI partial pipelines for
+            a set of data sources and sinks.
+            See the example below for the expected structure.
+        client_iri: IRI of OTELib client to use.
+
+    Returns:
+        Dict-representation of a declarative ExecFlow pipeline.
+
+    Examples:
+        Example of the `resources` argument:
+
+        ```python
+        {'sinks': [
+           {'iri': 'ss3:AbaqusConfiguration',
+            'resourcetype': 'ss3:AbaqusSimulation'},
+           {'iri': 'ss3:AluminiumMaterialCard',
+            'resourcetype': 'ss3:AbaqusSimulation'},
+           {'iri': 'ss3:ConcreteMaterialCard',
+            'resourcetype': 'ss3:AbaqusSimulation'}],
+         'sources': [
+           {'iri': 'ss3kb:abaqus_config1', 'resourcetype': 'dataset'},
+           {'iri': 'ss3:AluminiumMaterialCard',
+            'resourcetype': 'ss3:MaterialCardGenerator'},
+           {'iri': 'ss3kb:abaqus_materialcard_concrete1',
+            'resourcetype': 'dataset'}]}
+        ```
+    """
+    names = {"input": [], "output": []}
+    strategies = []
+
+    lst = [("output", s) for s in resources.get("sources", [])]
+    lst.extend([("input", s) for s in resources.get("sinks", [])])
+    i = 1
+    for dtype, dct in lst:
+        iri = dct["iri"]
+        resourcetype = dct["resourcetype"]
+        suffix = (
+            iri.split("#", 1)[-1] if "#" in iri else iri.rsplit("/", 1)[-1]
+        )
+        if resourcetype == "dataset":
+            resource = load_container(ts, iri, recognised_keys=recognised_keys)
+        else:
+            r = load_simulation_resource(ts, resourcetype)
+            try:
+                resource = r[dtype][iri]
+            except KeyError:
+                resource = r[dtype][ts.prefix_iri(iri)]
+        for strategy in resource:
+            for stype, conf in strategy.items():
+                name = f"{suffix}_{stype}_{i}"
+                conf[stype] = name
+                i += 1
+                names[dtype].append(name)
+                strategies.append(conf)
+
+    strategies, names = add_execflow_decoration_to_pipeline(strategies, names)
+
+    return {
+        "version": 1,
+        "strategies": strategies,
+        "pipelines": {
+            "pipe": " | ".join(names["output"])
+            + " | "
+            + " | ".join(names["input"])
+        },
+    }
+
+
+def add_execflow_decoration_to_pipeline(strategies, names):
+    """Add ExecFlow decoration to the pipeline.
+
+    Arguments:
+        strategies: List of strategies in the pipeline.
+        names: Dict with one list of names for the output (source)
+          strategies and one list of names for the input (sink) strategies.
+    """
+    functions = [f for f in strategies if "function" in f.keys()]
+
+    # Add datanode2cuds if not already present, if a function is present
+    # in the source strategies.
```

Comment:
I might not be right, but is this really the logic's we want to implement? It assumes that no output source has a function strategy. But I think that we should be allowed to include a function strategy in an output source if it requires a known transformation.

Isn't the logics that we simply want to append a datanode2cuds strategy to each output source and a cuds2datanode to each input sink?

Instead of postprocessing the strategies/names lists, I think that this is easier done in the outer loop of generate_ontoflow_pipeline(). Having a help function to add the strategy is a very good idea.

Reply:
datanode2cuds should be prepended the source partial pipelines  if there is a function there because this means that there is an aiida datanode passed to oteapi from ExecFlow. We cannot always have it, only when there is a "to_cuds" with a list of labels in the workflow step (in the Declaratice Chain) calling oteapi. 

Similarly, the other way, but with the additional complication of making specific strategies for the files that should be passed over.

Note, however, that this step requires careful alignment with the creation of the Declaratice chain, and I think a reimplementation might be in order together with that.

Reply:
Lets take this in a separate PR. 

I think that the easiest would be that the `parse_ontoflow()` returns a simple data structure. For simplicity, I will suggest a list of pre-defined tuples, like `(iri, iotype, resourcetype, strategylist)`.